File: src/main/java/io/github/ajoz/workshop/fp/java/part_4/solutions/practice_1/Practice1.java
package io.github.ajoz.workshop.fp.java.part_4.solutions.practice_1;

import io.github.ajoz.workshop.fp.java.tools.CheckedSupplier;
import io.github.ajoz.workshop.fp.java.tools.Maybe;
import io.github.ajoz.workshop.fp.java.tools.Try;

import java.util.concurrent.atomic.AtomicReference;

@SuppressWarnings("unused")
interface Supplier<A> {
    A get();

    static <A> Supplier<A> ofChecked(final CheckedSupplier<A> supplier) {
        return () -> {
            try {
                return supplier.get();
            } catch (final Exception e) {
                throw new RuntimeException(e);
            }
        };
    }

    default Try<A> tryGet() {
        try {
            return Try.success(get());
        } catch (final Exception e) {
            return Try.failure(e);
        }
    }

    default Maybe<A> maybeGet() {
        try {
            return Maybe.some(get());
        } catch (final Exception e) {
            return Maybe.none();
        }
    }

    default A getOrElse(final A defaultValue) {
        try {
            return get();
        } catch (final Exception exception) {
            return defaultValue;
        }
    }

    default Supplier<A> memoized() {
        final AtomicReference<A> value = new AtomicReference<>();
        return () -> {
            synchronized (value) {
                if (value.get() == null) {
                    value.set(get());
                }
                return value.get();
            }
        };
    }
}

class SomethingHappenedToFoo extends Exception {
    SomethingHappenedToFoo() {
        super("Something very very very bad happened!");
    }
}

@SuppressWarnings({"WeakerAccess", "unused"})
public class Practice1 {
    public static String foo() throws Exception {
        throw new SomethingHappenedToFoo();
    }

    public static String foo2() throws SomethingHappenedToFoo {
        throw new SomethingHappenedToFoo();
    }

    private static final CheckedSupplier<String> checkedSupplier = Practice1::foo;
    private static final Supplier<String> supplier = Supplier.ofChecked(Practice1::foo2);

    public static void main(final String[] args) {
        try {
            checkedSupplier.get();
        } catch (Exception e) {
            System.out.println(String.format("Exception thrown: %s", e.getMessage()));
        }

        supplier.maybeGet()
                .ifSome(System.out::println)
                .ifNone(() -> System.out.println("none result :("));

        supplier.tryGet()
                .ifSuccess(System.out::println)
                .ifFailure(System.out::println);

        System.out.println(supplier.getOrElse("Value in case of error!"));

        final Supplier<String> memoized = supplier.memoized();
        for (int i = 0; i < 100; i++) {
            supplier.maybeGet()
                    .ifSome(System.out::println)
                    .ifNone(() -> System.out.println("none result :("));

            supplier.tryGet()
                    .ifSuccess(System.out::println)
                    .ifFailure(System.out::println);
        }
    }
}

File: src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/part_4/solutions/practice_1/Practice1.kt
@file:Suppress("PackageName")

package io.github.ajoz.workshop.fp.kotlin.part_4.solutions.practice_1

import io.github.ajoz.workshop.fp.kotlin.tools.control.Maybe
import io.github.ajoz.workshop.fp.kotlin.tools.control.Try
import java.util.concurrent.atomic.AtomicReference

internal fun <A> (() -> A).tryGet(): Try<A> =
        Try.of(this)

internal fun <A> (() -> A).maybeGet(): Maybe<A> = try {
    Maybe.Some(invoke())
} catch (exception: Exception) {
    Maybe.None()
}

internal fun <A> (() -> A).getOrElse(default: A): A = try {
    invoke()
} catch (exception: Exception) {
    default
}

internal fun <A> (() -> A).memoized(): () -> A {
    val memo: AtomicReference<A> = AtomicReference()
    return {
        synchronized(memo) {
            if (memo.get() == null)
                memo.set(invoke())
            memo.get()
        }
    }
}

internal class SomethingHappenedToFoo : Exception("Something very very very bad happened!")

@Throws(Exception::class)
fun foo(): String {
    throw SomethingHappenedToFoo()
}

fun main(args: Array<String>) {
    val supplier = ::foo

    try {
        supplier()
    } catch (e: Exception) {
        println(String.format("Exception thrown: %s", e.message))
    }

    supplier.maybeGet()
            .ifSome { println(it) }
            .ifNone { println("none result :(") }

    supplier.tryGet()
            .ifSuccess { println(it) }
            .ifFailure { println(it) }

    println(supplier.getOrElse("Value in case of error!"))

    val memoized = supplier.memoized()
    for (i in 0..9) {
        memoized.maybeGet()
                .ifSome { println(it) }
                .ifNone { println("none result :(") }

        memoized.tryGet()
                .ifSuccess { println(it) }
                .ifFailure { println(it) }
    }
}
